runtime/src/accounts_index.rs

```diff
@@ -181,9 +201,172 @@ impl<'a, T: 'static + Clone> Iterator for AccountsIndexIterator<'a, T> {
+                if let dashmap::mapref::entry::Entry::Occupied(inner_key_entry) =
+                    inner_key_map.entry(*inner_key)
+                {
+                    // Delete the `inner_key` if the slot set is empty
+                    let slot_set = inner_key_entry.get();
+                    if slot_set.is_empty() {
+                        inner_key_entry.remove();
+                    }
+                }
+                inner_key_map.is_empty()
+            } else {
+                false
+            };
+            is_key_empty
+        } else {
+            false
+        };
+
+        // Delete the `key` if the set of inner keys is epty
+        if is_key_empty {
+            if let dashmap::mapref::entry::Entry::Occupied(key_entry) = self.index.entry(*key) {
+                if key_entry.get().is_empty() {
+                    key_entry.remove();
+                }
+            }
+        }
+    }
+
+    // Specifying `slots_to_remove` == None will only remove keys for those specific slots
+    // found for the `inner_key` in the reverse index. Otherwise, passing `None`
+    // will  remove all keys that are found for the `inner_key` in the reverse index.
+
+    // Note passing `None` is dangerous unless you're sure there's no other competing threads
+    // writing updates to the index for this Pubkey at the same time!
+    fn remove_by_inner_key(&self, inner_key: &Pubkey, slots_to_remove: Option<&HashSet<Slot>>) {
+        // Save off which keys in `self.index` had slots removed so we can remove them
+        // after we purge the reverse index
+        let mut key_to_removed_slots: HashMap<Pubkey, Vec<Slot>> = HashMap::new();
+
+        // Check if the entry for `inner_key` in the reverse index is empty
+        // and can be removed
+        let needs_remove = {
+            if let Some(slots_to_remove) = slots_to_remove {
+                self.reverse_index
+                    .get(inner_key)
+                    .map(|slots_map| {
+                        for slot in slots_to_remove.iter() {
+                            if let Some(removed) = slots_map.value().remove(slot) {
+                                key_to_removed_slots
+                                    .entry(removed.1)
+                                    .or_default()
+                                    .push(*slot);
+                            }
+                        }
+                        slots_map.is_empty()
+                    })
+                    .unwrap_or(false)
+            } else {
+                if let Some((_, removed_slot_map)) = self.reverse_index.remove(inner_key) {
+                    for (slot, key) in removed_slot_map {
+                        key_to_removed_slots.entry(key).or_default().push(slot);
+                    }
+                }
+                // We just removed the key, no need to remove it again
+                false
+            }
+        };
+
+        if needs_remove {
```

Comment:
@ryoqun would be good to have you  double check this type of delete only if empty logic, it shows up quite a few times in this PR, sorry!

The get called from `purge`, `purge_exact()` and `clean_rooted_entries`. The write lock for the account pubkey in the AccountsIndex is always held for these, so in the mapping

Mint pubkey -> Account pubkey -> Slots, the operations on the account and slot clearing should be safe, the only concurrent access is on the Mint pubkey, but I think the double checking empty before removing here: https://github.com/solana-labs/solana/blob/51ebca2a6cd2e17879d7bbae28a2250746c356fd/runtime/src/accounts_index.rs#L289-L290 should cover that as well

Reply:
it looks like this if body isn't tested at all?: https://codecov.io/gh/solana-labs/solana/pull/14212/diff#D1-376 I'd like a one test at least. :)

![image](https://user-images.githubusercontent.com/117807/103093780-b09e1980-463e-11eb-8e47-aea0be9915df.png)

Reply:
> Mint pubkey -> Account pubkey -> Slots, the operations on the account and slot clearing should be safe, the only concurrent access is on the Mint pubkey, but I think the double checking empty before removing here:
> 
> https://github.com/solana-labs/solana/blob/51ebca2a6cd2e17879d7bbae28a2250746c356fd/runtime/src/accounts_index.rs#L289-L290
> should cover that as well

Yeah, I think so too.

Reply:
:eyes:

Reply:
@ryoqun fixed a bug and added tests that should cover this case now: https://github.com/solana-labs/solana/pull/14212/commits/5a46407e73480eed5039d87ef1ae638bbcf6dd77